vortex-sampling-compressor/src/compressors/chunked.rs

```diff
@@ -0,0 +1,142 @@
+        None => return Ok(None),
+        Some(tree) => tree.into_parts(),
+    };
+
+    let Some(target_ratio) = metadata else {
+        vortex_bail!("chunked array compression tree must have metadata")
+    };
+
+    let Some(ChunkedCompressorMetadata(target_ratio)) =
+        target_ratio.as_ref().as_any().downcast_ref()
+    else {
+        vortex_bail!(
+            "chunked array compression
+        tree must ChunkedCompressorMetadata"
+        )
+    };
+
+    if children.len() != 1 {
+        vortex_bail!("chunked array compression tree must have one child")
+    }
+
+    let child = children.remove(0);
+
+    match (child, target_ratio) {
+        (None, None) => Ok(None),
+        (Some(child), Some(ratio)) => Ok(Some((child, *ratio))),
+        (..) => vortex_bail!("chunked array compression tree must have a child iff it has a ratio"),
+    }
+}
+
+impl ChunkedCompressor {
+    fn compress_chunked<'a>(
+        &'a self,
+        array: &ChunkedArray,
+        mut previous: Option<(CompressionTree<'a>, f32)>,
+        ctx: SamplingCompressor<'a>,
+    ) -> VortexResult<CompressedArray<'a>> {
+        let less_chunked = array.rechunk(
+            ctx.options().target_block_bytesize,
+            ctx.options().target_block_size,
+        )?;
+        let mut compressed_chunks = Vec::with_capacity(less_chunked.nchunks());
+        for (index, chunk) in less_chunked.chunks().enumerate() {
+            let like = previous.as_ref().map(|(like, _)| like);
+            let (compressed_chunk, tree) = ctx
+                .named(&format!("chunk-{}", index))
+                .compress(&chunk, like)?
+                .into_parts();
+
+            let ratio = (compressed_chunk.nbytes() as f32) / (chunk.nbytes() as f32);
+            let exceeded_target_ratio = previous
+                .as_ref()
+                .map(|(_, target_ratio)| ratio > target_ratio * 1.2)
+                .unwrap_or(false);
+
+            if ratio > 1.0 || exceeded_target_ratio {
+                warn!("unsatisfactory ratio {} {:?}", ratio, previous);
+                let (compressed_chunk, tree) = ctx.compress_array(&chunk)?.into_parts();
+                let new_ratio = (compressed_chunk.nbytes() as f32) / (chunk.nbytes() as f32);
+                previous = tree.map(|tree| (tree, new_ratio));
+                compressed_chunks.push(compressed_chunk);
+            } else {
+                previous = previous.or_else(|| tree.map(|tree| (tree, ratio)));
+                compressed_chunks.push(compressed_chunk);
+            }
+        }
+
+        let (child, ratio) = match previous {
+            Some((child, ratio)) => (Some(child), Some(ratio)),
+            None => (None, None),
+        };
+
+        Ok(CompressedArray::new(
+            ChunkedArray::try_new(compressed_chunks, array.dtype().clone())?.into_array(),
+            Some(CompressionTree::new_with_metadata(
+                self,
+                vec![child],
```

Comment:
if child is None, should this just be empty...?

Reply:
FWIW, this is the diff. I don't find either of these particularly palatable but I think the fix is to sort out how compressors pass information from one invocation of compress to the next.

```diff
(docs) # g diff
diff --git a/vortex-sampling-compressor/src/compressors/chunked.rs b/vortex-sampling-compressor/src/compressors/chunked.rs
index c0f5aebe..9adf4ab4 100644
--- a/vortex-sampling-compressor/src/compressors/chunked.rs
+++ b/vortex-sampling-compressor/src/compressors/chunked.rs
@@ -78,16 +78,16 @@ fn like_into_parts(
         vortex_bail!("chunked array compression tree must be ChunkedCompressorMetadata")
     };
 
-    if children.len() != 1 {
-        vortex_bail!("chunked array compression tree must have one child")
+    if (children.len() == 1) != target_ratio.is_some() {
+        vortex_bail!("chunked array compression tree must have a child iff it has a ratio")
     }
 
-    let child = children.remove(0);
-
-    match (child, target_ratio) {
-        (None, None) => Ok(None),
-        (Some(child), Some(ratio)) => Ok(Some((child, *ratio))),
-        (..) => vortex_bail!("chunked array compression tree must have a child iff it has a ratio"),
+    if children.len() == 0 {
+        return Ok(None);
+    } else if children.len() == 1 {
+        return Ok(Some((children.remove(0).unwrap(), target_ratio.unwrap())));
+    } else {
+        vortex_bail!("chunked array compression tree must have at most one child")
     }
 }
 
@@ -141,16 +141,16 @@ impl ChunkedCompressor {
             }
         }
 
-        let (child, ratio) = match previous {
-            Some((child, ratio)) => (Some(child), Some(ratio)),
-            None => (None, None),
+        let (children, ratio) = match previous {
+            Some((child, ratio)) => (vec![Some(child)], Some(ratio)),
+            None => (vec![], None),
         };
 
         Ok(CompressedArray::new(
             ChunkedArray::try_new(compressed_chunks, array.dtype().clone())?.into_array(),
             Some(CompressionTree::new_with_metadata(
                 self,
-                vec![child],
+                children,
                 Arc::new(ChunkedCompressorMetadata(ratio)),
             )),
         ))
```